app/CodeFormatterApp.php

```diff
@@ -32,59 +33,96 @@ public function __construct(Parser $parser = null)
      */
     public function run()
     {
-        $tokens = $this->parseInput();
+        $nodes = $this->parseInput();
 
-        foreach ($tokens as $token) {
-            echo $this->formatToken($token);
+        foreach ($nodes as $node) {
+            echo $this->formatNode($node);
         }
     }
 
     /**
      * parses the code from stdin.
      *
-     * @return iterable<Token>
+     * @return iterable<Node>
      */
     private function parseInput(): iterable
     {
         return $this->parser->parseFile('php://input');
     }
 
     /**
-     * formats a token based on its type and language.
+     * formats a node based on its type and language.
      *
-     * @param Token $token
+     * @param Node $node
      *
      * @return string
      */
-    private function formatToken(Token $token): string
+    private function formatNode(Node $node): string
     {
-        if ($token->isText()) {
-            return $token->getBody();
+        if ($node instanceof TextNode) {
+            return $this->exportNode($node, null);
         }
 
-        $language = $token->getAttribute('lang');
-        \assert(null !== $language);
+        \assert($node instanceof ElemNode);
 
+        if (!$node->isCode()) {
+            // export node as-is (PLAIN bbcode)
+            return $this->exportNode($node, null);
+        }
+
+        $nodeBody = $node->getBody();
+
+        if (\is_array($nodeBody)) {
+            $buffer = '';
+            foreach ($nodeBody as $subNode) {
+                $buffer .= $this->formatNode($subNode);
+            }
+            return $this->exportNode($node, $buffer);
```

Comment:
I'm not sure if this is an issue, but CODE=rich skips the formatting part. 
We can change this, but i currently see no reason to format "rich" code (nested code-nodes get formatted as before).

Reply:
No, this is totally fine. 😃